internal/driver/driver.go

```diff
@@ -0,0 +1,646 @@
+	}
+	d.activeDevices[deviceName] = activeDevice
+	d.lc.Debugf("a new Device is added: %s", deviceName)
+	if activeDevice.autoStreaming {
+		edgexErr = d.startStreaming(activeDevice)
+		if edgexErr != nil {
+			return errors.NewCommonEdgeXWrapper(edgexErr)
+		}
+	}
+	return nil
+}
+
+// UpdateDevice is a callback function that is invoked
+// when a Device associated with this Device Service is updated
+func (d *Driver) UpdateDevice(deviceName string, protocols map[string]models.ProtocolProperties,
+	adminState models.AdminState) error {
+	if edgexErr := d.RemoveDevice(deviceName, protocols); edgexErr != nil {
+		return errors.NewCommonEdgeXWrapper(edgexErr)
+	}
+	if edgexErr := d.AddDevice(deviceName, protocols, adminState); edgexErr != nil {
+		return errors.NewCommonEdgeXWrapper(edgexErr)
+	}
+	d.lc.Debugf("Device %s is updated", deviceName)
+	return nil
+}
+
+// RemoveDevice is a callback function that is invoked
+// when a Device associated with this Device Service is removed
+func (d *Driver) RemoveDevice(deviceName string, protocols map[string]models.ProtocolProperties) error {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	if device, ok := d.activeDevices[deviceName]; ok {
+		device.StopStreaming()
+		delete(d.activeDevices, deviceName)
+		d.lc.Debugf("Device %s is removed", deviceName)
+	}
+	return nil
+}
+
+// RefreshExistingDevicePaths checks whether the existing devices match the connected devices.
+// If there is a mismatch between them, scan all paths to find the matching device and update the existing device with the correct path.
+func (d *Driver) RefreshExistingDevicePaths() {
+	for _, cd := range d.ds.Devices() {
+		fdPath := cd.Protocols[UsbProtocol][Path]
+		cn, sn, err := getUSBDeviceIdInfo(fdPath)
+		if err != nil {
+			d.lc.Errorf("failed to get the serial number of device %s, error: %s", cd.Name, err.Error())
+		}
+		// If the card name or serial number is different, it means that the path of the device has changed.
+		if cn != cd.Protocols[UsbProtocol][CardName] || sn != cd.Protocols[UsbProtocol][SerialNumber] {
+			go d.updateDevicePath(cd)
+		}
+	}
+}
+
+// Discover triggers protocol specific device discovery, which is an asynchronous operation.
+// Devices found as part of this discovery operation are written to the channel devices.
+func (d *Driver) Discover() {
+	var devices []sdkModels.DiscoveredDevice
+	// Convert the slice of cached devices to map in order to improve the performance in the subsequent for loop.
+	cdm := d.cachedDeviceMap()
+	// The file descriptor of video capture device can be /dev/video0 ~ 63
+	// https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/devices.txt#L1402-L1406
+	for i := 0; i < 64; i++ {
+		fdPath := BasePath + strconv.Itoa(i)
+		if ok := d.isVideoCaptureDevice(fdPath); ok {
+			cn, sn, err := getUSBDeviceIdInfo(fdPath)
+			if err != nil {
+				d.lc.Errorf("failed to get device serial number, error: %s", err.Error())
+				continue
+			}
+			if _, ok := cdm[cn+sn]; ok {
```

Comment:
change variable name so it has meaning so easer to understand the code.
```suggestion
			if _, ok := currentDevices[cn+sn]; ok {
```

Reply:
Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/b915490962f8488701e51706647b02e88983813a